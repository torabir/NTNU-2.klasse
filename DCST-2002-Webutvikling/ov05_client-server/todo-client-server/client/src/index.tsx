import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Component } from 'react-simplified';
import { Card, Row, Column, Form, Button } from './widgets';
import taskService, { Task } from './task-service';

class TaskList extends Component {
  tasks: Task[] = [];

  // A: 
  // Legger til en metode for å oppdatere den gitte task basert på id når checkbox endres
  toggleTaskDone(taskId: number) {
    this.tasks = this.tasks.map((task) => {
      if (task.id === taskId) {
        const updatedTask = {
          ...task,
          done: !task.done
        };
  
        console.log('Updating task:', updatedTask);  // Legg til logging her for å se at riktig data sendes
  
        taskService.update(updatedTask)
          .then(() => {
            console.info(`Task ${taskId} successfully updated on server`);
          })
          .catch((error) => {
            console.error(`Failed to update task ${taskId} on server:`, error);
          });
  
        return updatedTask;
      } else {
        return task;
      }
    });
  
    this.setState({ tasks: this.tasks });
  }
  
  // toggleTaskDone(taskId: number) {
  //   this.tasks = this.tasks.map((task) => {
  //     if (task.id === taskId) {
  //       const updatedTask = {
  //         ...task,              // Kopierer alle egenskaper fra task
  //         done: !task.done      // Toggler done-status
  //       };
  
  //       taskService.update(updatedTask)  // Sender hele updatedTask-objektet
  //         .then(() => {
  //           console.info(`Task ${taskId} successfully updated on server`);
  //         })
  //         .catch((error) => {
  //           console.error(`Failed to update task ${taskId} on server:`, error);
  //         });
  
  //       return updatedTask;
  //     } else {
  //       return task;
  //     }
  //   });
  
  //   this.setState({ tasks: this.tasks });
  // }
  

  // B: Metode for å slette en oppgave
  deleteTask(taskId: number) {
    taskService.delete(taskId)
      .then(() => {
        this.tasks = this.tasks.filter((task) => task.id !== taskId);
        this.setState({ tasks: this.tasks });
        console.log(`Task ${taskId} successfully deleted.`);
      })
      .catch((error) => {
        console.error(`Failed to delete task ${taskId}:`, error);
      });
  }

  render() {
    return (
      <Card title="Tasks">
        {/* A: Legger til 1 rad med 3 kolonner for overskrifter */}
        <Row>
          <Column><strong>Title</strong></Column>
          <Column><strong>Done</strong></Column>
          <Column><strong>Delete</strong></Column>
        </Row>
        {this.tasks.map((task) => (
          <Row key={task.id}>
            <Column>{task.title}</Column>
            {/* B: Checkbox for å markere oppgaven som done/ikke done */}
            <Column>
              <input
                type="checkbox"
                checked={task.done}
                onChange={() => this.toggleTaskDone(task.id)}
              />
            </Column>
            <Column>
              <Button.Danger small onClick={() => this.deleteTask(task.id)}>X</Button.Danger>
            </Column>
          </Row>
        ))}
      </Card>
    );
  }

  // A: Når komponenten monteres, henter den alle oppgaver fra serveren
  mounted() {
    taskService.getAll().then((tasks) => {
      this.tasks = tasks; // Henter oppgavene fra serveren
      this.setState({ tasks: this.tasks }); // Oppdaterer state for å vise oppgavene i UI
    });
  }
}

class TaskNew extends Component {
  title = '';

  render() {
    return (
      <Card title="New task">
        <Row>
          <Column width={1}>
            <Form.Label>Title:</Form.Label>
          </Column>
          <Column width={4}>
            <Form.Input
              type="text"
              value={this.title}
              onChange={(event) => (this.title = event.currentTarget.value)}
            />
          </Column>
        </Row>
        <Button.Success
          onClick={() => {
            taskService.create(this.title).then(() => {
              // Reloads the tasks in the Tasks component
              TaskList.instance()?.mounted(); // .? = call TaskList.instance().mounted() if TaskList.instance() does not return null
              this.title = ''; // Tømmer inputfeltet etter opprettelse
            });
          }}
        >
          Create
        </Button.Success>
      </Card>
    );
  }
}

let root = document.getElementById('root');
if (root) {
  createRoot(root).render(
    <>
      <TaskList />
      <TaskNew />
    </>
  );
}
