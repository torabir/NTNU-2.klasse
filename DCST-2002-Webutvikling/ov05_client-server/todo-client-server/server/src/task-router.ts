import express from 'express';
import taskService from './task-service';

/**
 * Express router containing task methods.
 */
const router = express.Router();

// Henter alle oppgaver
router.get('/tasks', (_request, response) => {
  taskService
    .getAll() // Kaller getAll service for å hente alle oppgaver
    .then((rows) => response.send(rows)) // Sender alle tasks som svar
    .catch((error) => response.status(500).send(error)); // Feilhåndtering
});

// Henter en spesifikk oppgave ved ID
router.get('/tasks/:id', (request, response) => {
  const id = Number(request.params.id); // Konverterer ID fra URL til et tall
  taskService
    .get(id) // Kaller get service for å hente oppgave med gitt ID
    .then((task) => (task ? response.send(task) : response.status(404).send('Task not found'))) // Sender tasken eller 404 hvis tasken ikke finnes
    .catch((error) => response.status(500).send(error)); // Feilhåndtering
});

// Oppretter en ny oppgave
router.post('/tasks', (request, response) => {
  const data = request.body; // Henter oppgave-data fra forespørselen
  if (data && data.title && data.title.length != 0) { // Sjekker at tittelen ikke er tom
    taskService
      .create(data.title) // Kaller create service for å opprette oppgaven
      .then((id) => response.send({ id: id })) // Sender tilbake ID-en til den opprettede tasken
      .catch((error) => response.status(500).send(error)); // Feilhåndtering
  } else {
    response.status(400).send('Missing task title'); // Returnerer feil hvis tittelen mangler
  }
});

// Sletter en oppgave med gitt ID
router.delete('/tasks/:id', (request, response) => {
  taskService
    .delete(Number(request.params.id)) // Kaller delete service for å slette oppgaven med gitt ID
    .then((_result) => response.sendStatus(200)) // Sender svar hvis sletting er vellykket
    .catch((error) => response.status(500).send(error)); // Feilhåndtering
});

      // A
/**
 * PUT-rute for å oppdatere oppgave basert på ID.
 * Forventet request body: { done: true/false }
 */ 

router.put('/tasks/:id', (request, response) => {
  const id = Number(request.params.id);
  const { done } = request.body;

  console.log('Received PUT request for task:', { id, done });  // Legg til logging her for å bekrefte dataen

  if (typeof done === 'boolean') {
    taskService
      .get(id)
      .then((task) => {
        if (!task) return response.status(404).send('Task not found');

        const updatedTask = { ...task, done: done };

        taskService.update(updatedTask)
          .then(() => response.sendStatus(200))
          .catch((error) => response.status(500).send(error));
      })
      .catch((error) => response.status(500).send(error));
  } else {
    response.status(400).send('Invalid data');
  }
});


// GAMMEL: 
// router.put('/tasks/:id', (request, response) => {
//   const id = Number(request.params.id);
//   const task = request.body;

//   taskService
//     .update(id, task)
//     .then(() => response.sendStatus(200))
//     .catch((error) => response.status(500).send(error));
// });

export default router;
