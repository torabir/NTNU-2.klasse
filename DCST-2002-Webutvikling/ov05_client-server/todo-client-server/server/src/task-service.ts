import pool from './mysql-pool';
import type { RowDataPacket, ResultSetHeader } from 'mysql2';

export type Task = {
  id: number;
  title: string;
  done: boolean;
  description: string; 
};

class TaskService {
  /**
   * Get task with given id.
   */
  get(id: number) {
    return new Promise<Task | undefined>((resolve, reject) => { // det inni <> er typen som returneres av Promiset (resolve)
      pool.query('SELECT * FROM Tasks WHERE id = ?', [id], (error, results: RowDataPacket[]) => {
        if (error) return reject(error);

        resolve(results[0] as Task); // Returnerer den første tasken som matcher ID-en
      });
    });
  }

  /**
   * Get all tasks.
   */
  getAll() {
    return new Promise<Task[]>((resolve, reject) => {
      pool.query('SELECT * FROM Tasks', [], (error, results: RowDataPacket[]) => {
        if (error) return reject(error);

        resolve(results as Task[]); // Promiset returnerer en liste med Tasks
      });
    });
  }

  /**
   * Create new task having the given title.
   *
   * Resolves the newly created task id.
   */
  create(title: string) {
    return new Promise<number>((resolve, reject) => {
      pool.query('INSERT INTO Tasks SET title=?', [title], (error, results: ResultSetHeader) => {
        if (error) return reject(error);

        resolve(results.insertId); // Promiset returnerer et tall (ID-en til den opprettede oppgaven)
      });
    });
  }

  /**
   * Delete task with given id.
   */
  delete(id: number) {
    return new Promise<void>((resolve, reject) => {
      pool.query('DELETE FROM Tasks WHERE id = ?', [id], (error, results: ResultSetHeader) => {
        if (error) return reject(error);
        if (results.affectedRows == 0) reject(new Error('No row deleted'));

        resolve(); // Bekrefter at slettingen var vellykket
      });
    });
  }

  // A:
  /**
   * Oppdaterer en oppgave i databasen.
   */
  update(task: Task) {
    return new Promise<void>((resolve, reject) => {
      pool.query(
        'UPDATE Tasks SET done = ? WHERE id = ?',  // Kun oppdaterer `done`-feltet i denne spørringen
        [task.done, task.id],
        (error, results: ResultSetHeader) => {
          if (error) {
            console.error('SQL update error:', error);  // Her logger vi SQL-feilen
            return reject(error);  // Returner feilen
          }
          if (results.affectedRows === 0) {
            return reject(new Error('No task found to update'));  // Hvis ingen rader er oppdatert
          }
          resolve();  // Hvis oppdateringen er vellykket
        }
      );
    });
  }
  
  
  // Gammel: 
  // update(task: Task) {
  //   return new Promise<void>((resolve, reject) => {
  //     pool.query(
  //       'UPDATE tasks SET title = ?, done = ?, description = ? WHERE id = ?', [task.title, task.done, task.description, task.id],
  //       (error, results: ResultSetHeader) => {
  //         if (error) return reject(error);
  //         if (results.affectedRows === 0) return reject(new Error('No task found to update'));
  //         resolve();
  //       }
  //     );
  //   });
  // }
}

const taskService = new TaskService();
export default taskService;
