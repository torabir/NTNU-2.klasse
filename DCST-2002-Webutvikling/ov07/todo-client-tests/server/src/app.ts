import express from 'express';
import router from './task-router';

/**
 * Express application.
 */
const app = express();

app.use(express.json());

// Bruker API-versjon 2 for det nye endepunktet
app.use('/api/v2', router); // Endret fra TaskRouter til router

export default app;