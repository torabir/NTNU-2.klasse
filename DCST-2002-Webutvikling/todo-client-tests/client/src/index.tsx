import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Form, Button } from './widgets';
import axios from 'axios';

// Sett opp standard baseURL for axios (API-kommunikasjon)
axios.defaults.baseURL = 'http://localhost:3000/api/v2';

class Home extends Component {
  source = '';  // Koden som brukeren skriver
  exitStatus: number | null = null;
  stdout = '';  // Output fra koden
  stderr = '';  // Error fra koden

  render() {
    return (
      <>
        <Card title="app.js">
          <Form.Textarea
            value={this.source}
            onChange={(event) => (this.source = event.currentTarget.value)}
            rows={4}
          />
          <Button.Success
            onClick={() => {
              axios
                .post<{ exitStatus: number; stdout: string; stderr: string }>('/run', {
                  language: 'js',
                  source: this.source,
                })
                .then((response) => {
                  this.exitStatus = response.data.exitStatus;
                  this.stdout = response.data.stdout;
                  this.stderr = response.data.stderr;
                })
                .catch((error: Error) => Alert.danger('Could not run app.js: ' + error.message));
            }}
          >
            Run
          </Button.Success>
        </Card>
        <Card title="Standard output">{this.stdout}</Card>
        <Card title="Standard error">{this.stderr}</Card>
        <Card title={'Exit status: ' + (this.exitStatus ?? '')}></Card>
      </>
    );
  }
}

const root = document.getElementById('root');
if (root)
  ReactDOM.render(
    <>
      <Alert />
      <Home />
    </>,
    root
  );
