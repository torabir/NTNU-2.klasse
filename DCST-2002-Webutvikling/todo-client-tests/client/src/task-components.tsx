import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Row, Column, Form, Button } from './widgets';
import { NavLink } from 'react-router-dom';
import taskService, { Task } from './task-service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student

/**
 * Renders task list.
 */
export class TaskList extends Component {
  tasks: Task[] = [];

  render() {
    return (
      <>
        <Card title="Tasks">
          {this.tasks.map((task) => (
            <Row key={task.id}>
              <Column>
                <NavLink to={'/tasks/' + task.id}>{task.title}</NavLink>
              </Column>
            </Row>
          ))}
        </Card>
        <Button.Success onClick={() => history.push('/tasks/new')}>New task</Button.Success>
      </>
    );
  }

  mounted() {
    taskService
      .getAll()
      .then((tasks) => (this.tasks = tasks))
      .catch((error) => Alert.danger('Error getting tasks: ' + error.message));
  }
}

/**
 * Renders a specific task.
 */
export class TaskDetails extends Component<{ match: { params: { id: number } } }> {
  task: Task = { id: 0, title: '', done: false };

  render() {
    return (
      <>
        <Card title="Task">
          <Row>
            <Column width={2}>Title:</Column>
            <Column>{this.task.title}</Column>
          </Row>
          <Row>
            <Column width={2}>Description:</Column>
          </Row>
          <Row>
            <Column width={2}>Done:</Column>
            <Column>
              <Form.Checkbox checked={this.task.done} onChange={() => {}} disabled />
            </Column>
          </Row>
        </Card>
        <Button.Success
          onClick={() => history.push('/tasks/' + this.props.match.params.id + '/edit')}
        >
          Edit
        </Button.Success>
      </>
    );
  }

  mounted() {
    taskService
      .get(this.props.match.params.id)
      .then((task) => (this.task = task))
      .catch((error) => Alert.danger('Error getting task: ' + error.message));
  }
}

/**
 * Renders form to edit a specific task.
 */
export class TaskEdit extends Component<{ match: { params: { id: number } } }> {
  task: Task = { id: 0, title: '', done: false };

  render() {
    return (
      <>
        <Card title="Edit task">
          <Row>
            <Column width={2}>
              <Form.Label>Title:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.task.title}
                onChange={(event) => (this.task.title = event.currentTarget.value)}
              />
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Label>Description:</Form.Label>
            </Column>
            <Column>
              <Form.Textarea value="" onChange={() => {}} rows={10} disabled />
            </Column>
          </Row>
          <Row>
            <Column width={2}>Done:</Column>
            <Column>
              <Form.Checkbox
                checked={this.task.done}
                onChange={(event) => (this.task.done = event.currentTarget.checked)}
              />
            </Column>
          </Row>
        </Card>
        <Row>
          <Column>
            <Button.Success onClick={() => Alert.info('Not yet implemented')}>Save</Button.Success>
          </Column>
          <Column right>
            <Button.Danger onClick={() => Alert.info('Not yet implemented')}>Delete</Button.Danger>
          </Column>
        </Row>
      </>
    );
  }

  mounted() {
    taskService
      .get(this.props.match.params.id)
      .then((task) => (this.task = task))
      .catch((error) => Alert.danger('Error getting task: ' + error.message));
  }
}

/**
 * Renders form to create new task.
 */
export class TaskNew extends Component {
  title = '';

  render() {
    return (
      <>
        <Card title="New task">
          <Row>
            <Column width={2}>
              <Form.Label>Title:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.title}
                onChange={(event) => (this.title = event.currentTarget.value)}
              />
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Label>Description:</Form.Label>
            </Column>
            <Column>
              <Form.Textarea value="" onChange={() => {}} rows={10} disabled />
            </Column>
          </Row>
        </Card>
        <Button.Success
          onClick={() => {
            taskService
              .create(this.title)
              .then((id) => history.push('/tasks/' + id))
              .catch((error) => Alert.danger('Error creating task: ' + error.message));
          }}
        >
          Create
        </Button.Success>
      </>
    );
  }
}

// Ny class: 
// A: 
export class CodeRunner extends Component {
  code: string = ''; // Holder koden som brukeren skriver
  language: string = 'javascript'; // Standardvalget for språk

  output: string = ''; // Resultatet fra standard output
  error: string = ''; // Feilmeldinger fra standard error
  exitStatus: number = 0; // Exit status fra prosessen

  // Funksjon som sender koden og valgt språk til serveren for kjøring
  runCode() {
    axios
      .post('/api/run-code', { code: this.code, language: this.language })
      .then((response) => {
        this.output = response.data.output;
        this.error = response.data.error;
        this.exitStatus = response.data.exitStatus;
      })
      .catch((error) => {
        Alert.danger('Error running code: ' + error.message);
      });
  }

  render() {
    return (
      <Card title="Run Code">
        {/* Velg språk */}
        <Form.Select
          value={this.language}
          onChange={(event) => (this.language = event.currentTarget.value)}
        >
          <option value="javascript">JavaScript (Node.js)</option>
          <option value="python">Python</option>
          <option value="bash">Bash</option>
        </Form.Select>

        {/* Kode-innskrivning */}
        <Form.Textarea
          value={this.code}
          onChange={(event) => (this.code = event.currentTarget.value)}
          rows={10}
          placeholder="Write your code here..."
        />

        {/* Kjør kode */}
        <Button.Success onClick={() => this.runCode()}>Run</Button.Success>

        {/* Resultater */}
        <Card title="Results">
          <p><b>Standard Output:</b> {this.output}</p>
          <p><b>Standard Error:</b> {this.error}</p>
          <p><b>Exit Status:</b> {this.exitStatus}</p>
        </Card>
      </Card>
    );
  }
}