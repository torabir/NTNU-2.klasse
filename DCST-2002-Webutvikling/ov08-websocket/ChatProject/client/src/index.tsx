import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Whiteboard } from './whiteboard-component';
import { Alert } from './widgets';
import { Chat } from './chat-component';  

const root = document.getElementById('root');
if (root)
  createRoot(root).render(
    <>
      <Alert />
      {/* <Whiteboard /> */}
      <Chat />
    </>,
  );
