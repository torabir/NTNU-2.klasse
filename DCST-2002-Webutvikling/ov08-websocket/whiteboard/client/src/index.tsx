import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Whiteboard } from './whiteboard-component';
import { Alert } from './widgets';

let root = document.getElementById('root');
if (root)
  createRoot(root).render(
    <>
      <Alert />
      <Whiteboard />
    </>,
  );
