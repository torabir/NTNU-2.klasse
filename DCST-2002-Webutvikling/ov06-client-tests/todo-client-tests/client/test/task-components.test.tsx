import * as React from 'react';
import { TaskList, TaskNew, TaskDetails, TaskEdit } from '../src/task-components'; // lagt til TaskDetails, TaskEdit 
import * as Taskservice from '../src/task-service'; // lagt til . unødvendig fordi service mockes uansett. 
import { shallow } from 'enzyme';
import { Form, Button, Column } from '../src/widgets'; // Lagt til widgets
import { NavLink } from 'react-router-dom';


// jest.mock('../src/task-service', () => {
//   class TaskService {
//     getAll() {
//       return Promise.resolve([
//         { id: 1, title: 'Les leksjon', done: false },
//         { id: 2, title: 'Møt opp på forelesning', done: false },
//         { id: 3, title: 'Gjør øving', done: false },
//       ]);
//     }

//     create() {
//       return Promise.resolve(4); // Same as: return new Promise((resolve) => resolve(4));
//     }
//   }
//   return new TaskService();
// });

jest.mock('../src/task-service', () => {
  class TaskService {
    getAll() {
      return Promise.resolve([
        { id: 1, title: 'Les leksjon', description: 'Les kapittel 1', done: false },
        { id: 2, title: 'Møt opp på forelesning', description: 'Møt opp i rom A101', done: false },
        { id: 3, title: 'Gjør øving', description: 'Fullfør øving 1', done: false },
      ]);
    }

    get() {
      return Promise.resolve({ id: 1, title: 'Les leksjon', description: 'Les kapittel 1', done: false });
    }

    create() {
      return Promise.resolve(4); // Returnerer en ny oppgave-id (4)
    }

    update() {
      return Promise.resolve(); // Returnerer ingenting ved oppdatering
    }

    delete() {
      return Promise.resolve(); // Returnerer ingenting ved sletting
    }
  }

  return new TaskService();
});


describe('Task component tests', () => {
  test('TaskList draws correctly', (done) => {
    const wrapper = shallow(<TaskList />);

    // Wait for events to complete
    setTimeout(() => {
      expect(
        wrapper.containsAllMatchingElements([
          <NavLink to="/tasks/1">Les leksjon</NavLink>,
          <NavLink to="/tasks/2">Møt opp på forelesning</NavLink>,
          <NavLink to="/tasks/3">Gjør øving</NavLink>,
        ])
      ).toEqual(true);
      done();
    });
  });

  test('TaskNew correctly sets location on create', (done) => {
    const wrapper = shallow(<TaskNew />);

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Kaffepause' } });
    // @ts-ignore
    expect(wrapper.containsMatchingElement(<Form.Input value="Kaffepause" />)).toEqual(true);

    wrapper.find(Button.Success).simulate('click');
    // Wait for events to complete
    setTimeout(() => {
      expect(location.hash).toEqual('#/tasks/4');
      done();
    });
  });
});


// A: 
//**
// TaskDetails component tests: 

test('TaskDetails draws correctly', async () => {
  const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

  // Await mounted() and data fetching
  await new Promise((resolve) => setTimeout(resolve, 0));

  console.log(wrapper.debug()); // Denne vil vise hele strukturen til wrapperen

  expect(
    wrapper.containsAllMatchingElements([
      <Column>Les leksjon</Column>, // Forventet tittel
      <Column>Les kapittel 1</Column>, // Forventet beskrivelse
      <Form.Checkbox checked={false} disabled /> // Forventet "done" checkbox (ikke ferdig)
    ])
  ).toEqual(true);
});

//B 
//** */
// Snapshot: 

test('TaskDetails matches snapshot', async () => {
  const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

  // Await mounted() and data fetching
  await new Promise((resolve) => setTimeout(resolve, 0));

  // Sammenlign wrapper med snapshot
  expect(wrapper).toMatchSnapshot();
});

// C
//** */
// Oppnå 70% dekning for statements: 

// Test for TaskEdit-komponenten:

test('TaskEdit draws correctly and updates task', async () => {
  const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

  // Await mounted() and data fetching
  await new Promise((resolve) => setTimeout(resolve, 0));

  // Check if the form is rendered with correct values
  expect(wrapper.find(Form.Input).prop('value')).toEqual('Les leksjon');
  expect(wrapper.find(Form.Textarea).prop('value')).toEqual('Les kapittel 1');
  expect(wrapper.find(Form.Checkbox).prop('checked')).toEqual(false);

  // Simulate form changes
  wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Ny tittel' } });
  wrapper.find(Form.Textarea).simulate('change', { currentTarget: { value: 'Ny beskrivelse' } });
  wrapper.find(Form.Checkbox).simulate('change', { currentTarget: { checked: true } });

  // Check if the form reflects the new values
  expect(wrapper.containsMatchingElement(<Form.Input value="Ny tittel" />)).toEqual(true);
  expect(wrapper.containsMatchingElement(<Form.Textarea value="Ny beskrivelse" />)).toEqual(true);
  expect(wrapper.containsMatchingElement(<Form.Checkbox checked={true} />)).toEqual(true);

  // Simulate save button click
  wrapper.find(Button.Success).simulate('click');

  // Ensure that the update service is called
  await new Promise((resolve) => setTimeout(resolve, 0));
  expect(location.hash).toEqual('#/tasks/1');
});

// Test for sletting i TaskEdit-komponenten:

test('TaskEdit deletes task and navigates away', async () => {
  const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

  // Await mounted() and data fetching
  await new Promise((resolve) => setTimeout(resolve, 0));

  // Simulate delete button click
  wrapper.find(Button.Danger).simulate('click');

  // Ensure that the delete service is called and navigation happens
  await new Promise((resolve) => setTimeout(resolve, 0));
  expect(location.hash).toEqual('#/tasks');
});



// ERRORS: 


// Bruker ikke async/await
// test('TaskDetails draws correctly', (done) => {
//   const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

//   // Vent til "mounted" har kjørt og oppgaven er hentet
//   setTimeout(() => {
//     console.log(wrapper.debug()); // Denne vil vise hele strukturen til wrapperen

//     expect(
//       wrapper.containsAllMatchingElements([
//         <Column>Les leksjon</Column>, // Forventet tittel
//         <Column>Les kapittel 1</Column>, // Forventet beskrivelse
//         <Form.Checkbox checked={false} disabled /> // Forventet "done" checkbox (ikke ferdig)
//       ])
//     ).toEqual(true);
//     done();
//   });
// });

// Feilen her var at jeg bruker h1, p, osv. 
// describe('TaskDetails component tests', () => {
//   test('TaskDetails draws correctly', (done) => {
//     const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

//     // Vi venter på eventer for at mock-get-tjenesten skal returnere verdier
//     setTimeout(() => {
//       console.log(wrapper.html()); // Logg ut den faktiske renderen av komponenten
//       expect(
//         wrapper.containsAllMatchingElements([
//           <h1>Task Details</h1>,
//           <p>Title: Les leksjon</p>,
//           <p>Status: Not done</p>,
//         ])
//       ).toEqual(true);
//       done();
//     });
//   });
// });
