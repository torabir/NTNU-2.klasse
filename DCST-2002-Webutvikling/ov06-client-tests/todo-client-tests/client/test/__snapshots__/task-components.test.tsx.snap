// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskDetails matches snapshot 1`] = `
<Fragment>
  <Card
    title="Task"
  >
    <Row>
      <Column
        width={2}
      >
        Title:
      </Column>
      <Column>
        Les leksjon
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Description:
      </Column>
      <Column>
        Les kapittel 1
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          disabled={true}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <ButtonSuccess
    onClick={[Function]}
  >
    Edit
  </ButtonSuccess>
</Fragment>
`;
